lib/resources/oidc.js

```diff
@@ -0,0 +1,280 @@
+          body { margin:0; font-family:"Helvetica Neue", Helvetica, Arial, sans-serif; background-color:#f7f7f7; }
+          .header { background-color:#bd006b; color:white; box-shadow: 0 3px 0 #dedede; border-top: 3px solid #8d0050; padding:0.5em 0; }
+          .header a,.header a:active,.header a:visited { margin:1em; font-size:12px; font-weight:700; color:white; text-decoration:none; }
+          #content { margin:3em auto; width:80%; background-color:white; border-color:rgb(51, 51, 51); box-shadow:rgba(0, 0, 0, 0.25) 0px 0px 24px 0px, rgba(0, 0, 0, 0.28) 0px 35px 115px 0px; }
+          #content h1 { background-color:#bd006b; color:white; border-bottom:1px solid #ddd; padding:10px 15px; font-size:18px; margin:0; }
+          #content div { border-bottom:1px solid #ddd; padding:10px 15px; }
+          #content div:last-child { border-bottom:none; background-color:#eee; }
+          #content div:last-child a { background-color:#009ecc; color:white; display:inline-block; padding:6px 10px 5px; border-radius:2px; text-decoration:none; font-size:12px; border-color:#286090; }
+          #content div:last-child a:hover { background-color:#0086ad; border-color:#204d74; }
+        </style>
+      </head>
+      <body>
+        <div class="header"><a href="/">ODK Central</a></div>
+        <div id="content">
+          ${body}
+        </div>
+      </body>
+    </html>
+  `;
+}
+
+function userNotFound(res, email) {
+  return clientError(res, `Authentication successful, but there is no user in the system with the supplied email address (${email}).`); // TODO reject unauthorized?  or don`t leak info?  etc.
+}
+
+function emailNotVerified(res, userinfo) {
+  // TODO what should this do?  redirect to login page with a toast informing the user something?
+
+  const { name, email } = userinfo;
+
+  return respond(res, 403, page(html`
+    <h1>Hello, ${name} (${email})!</h1>
+    <h2>Your email is not verified.</h2>
+    <h2>TODO</h2>
+    <div><pre>
+      * delete session and allow user to retry
+    </pre></div>
+    <div><a href="/">Go home</a></div>
+  `));
+}
+
+function debugIdpResponse(res, session, userinfo, user) {
+  const { name, email, picture } = userinfo;
+
+  return respond(res, page(html`
+    <h1>Hello, ${name} (${email})!</h1>
+    <div><h3>Profile picture</h3><img src="${picture}" style="min-width:200px; min-height:200px; border:solid black 1px"/></div>
+    <div><h3>User     </h3><code><pre>${JSON.stringify(user,     null, 2)}</pre></code></div>
+    <div><h3>Session  </h3><code><pre>${JSON.stringify(session,  null, 2)}</pre></code></div>
+    <div><h3>User Info</h3><code><pre>${JSON.stringify(userinfo, null, 2)}</pre></code></div>
+    <div>
+      <h2>TODO</h2>
+      <pre>
+        * stop displaying error details to client
+        * hide set-password UI
+        * don't request password from admin when creating new user account
+      </pre>
+    </div>
+    <div><a href="/">Continue to central</a></div>
+  `));
+}
```

Comment:
Frontend would also provide i18n for any text.

Reply:
:+1: This should be possible for errors if specific codes are defined and then a corresponding view is implemented in frontend.

For the success path, I think the extra backend step is required to maintain the current session/csrf cookie settings.

Reply:
> For the success path, I think the extra backend step is required to maintain the current session/csrf cookie settings.

Could you say more about why this is required? If this requirement is due to the setup in Frontend, maybe we could make changes there. Would it help to make the Frontend request to /v1/oidc/login async?

Reply:
Session ID and CSRF cookies are both `SameSite=Strict`:
https://github.com/getodk/central-backend/blob/facf33916322f4cea6e2b963372b4e7623715320/lib/resources/sessions.js#L39-L42
This means that if backend redirects to frontend at https://github.com/getodk/central-backend/blob/e7da23a3d3ac248f19a179e59a432c756f7c6361/lib/resources/oidc.js#L130-L146, the browser does not send cookies, so the user cannot be authorised.

There are tests for this at:

* https://github.com/getodk/central-backend/blob/e7da23a3d3ac248f19a179e59a432c756f7c6361/oidc-tester/playwright-tests/src/happy.spec.js
* https://github.com/getodk/central-backend/blob/e7da23a3d3ac248f19a179e59a432c756f7c6361/oidc-tester/playwright-tests/src/happy-next.spec.js

With a 302 redirect, the tests fail like this:

```
3) [firefox] › happy-next.spec.js:21:1 › can log in ──────────────────────────────────────────────

  AssertionError: No session cookie found!

     at utils.js:55

    53 |   console.log('requestCookies:', JSON.stringify(requestCookies, null, 2));
    54 |
  > 55 |   assert(requestCookies[SESSION_COOKIE], 'No session cookie found!');
       |   ^   
    56 |   assert(requestCookies['__csrf'],       'No CSRF cookie found!');
    57 |   assert.equal(Object.keys(requestCookies).length, 2, 'Unexpected requestCookie count!');
    58 | } 

      at assertLoginSuccessful (/odk-central-backend/oidc-tester/playwright-tests/src/utils.js:55:3)
      at /odk-central-backend/oidc-tester/playwright-tests/src/happy-next.spec.js:25:3
```

See e.g. https://github.com/alxndrsn/odk-central-backend/actions/runs/5666083718/job/15352099923#step:5:6265

Reply:
OK, that's super helpful, thank you! I think I'm understanding things better now. Just to confirm:

1. An error results in a redirect to Frontend. The redirect specifies an error code as a query parameter, which Frontend uses to show an alert.
2. A success renders an HTML page. The page does a client-side redirect to Frontend and also links to Frontend.

That all sounds good to me. :+1: For the success page, what do you think about leaving that as unstyled as possible and not trying to make it look like Frontend? Since the user will only be on the page for a short time and is automatically redirected, I don't feel like it needs to be styled. I've seen that pattern elsewhere and feel like it wouldn't be overly confusing to users.

Reply:
There's a parallel discussion at https://github.com/getodk/central-backend/pull/910/files#r1310995967:

I've also seen the pattern of unstyled forwarding pages elsewhere and I'm not a fan!  I think a branded forwarding page maintains users' trust that they are inside a legitimate workflow.

Reply:
I've filed a related issue: #971